File: Day 1/Advent of Code - Day 1.R
# Load required packages
require(dplyr)

# Load the dataset
setwd('~/Data Science/Advent of Code/2018/Day 1')
freq <- read.csv('frequency adjustment.csv')

# Part 1: Cumulative sum frequency
cumFreq <- cumsum(freq$frequency)
firstStar <- tail(cumFreq, n = 1)

# Turn to data frame
cumFreq <- data.frame('frequency' = cumFreq)

# Find all the numbers that repeat themselves
final <- data.frame()
while(nrow(final) == 0){
  lastOne <- tail(cumFreq, n = 1)
  freq2 <- lastOne %>% bind_rows(freq)
  cumFreq2 <- cumsum(freq2$frequency)
  cumFreq <- cumFreq %>% bind_rows(data.frame('frequency' = cumFreq2[-1])) %>% mutate(order = 1:n())
  final <- cumFreq %>% arrange(frequency, order) %>% group_by(frequency) %>% mutate(dedup = 1:n()) %>% filter(dedup == 2)
}

# Order the repeated numbers; first one is the answer
final %>% arrange(order) %>% head()

File: Day 1/Advent of Code - Day 1.py
# Advent of Code: Day 1
import os 
import numpy as np 
import pandas as pd
import itertools

# Load the dataset
os.chdir('C:/Users/malon/Documents/Data Science/Advent of Code/2018/Adventofcode2018/Day 1')
frequencyAdjustment = pd.read_csv('frequency adjustment.csv')

def day1_1(input: str):
    return sum(input)

day1_1(frequencyAdjustment['frequency'])

def day1_2(input: str):
    frequencies = set()
    currentFrequency = 0
    
    for i in itertools.cycle(input):
        currentFrequency += i

        oldLen = len(frequencies)
        frequencies.add(currentFrequency)
        newLen = len(frequencies)

        if oldLen == newLen:
            return currentFrequency

day1_2(frequencyAdjustment['frequency'])

